### PR TITLE
[FIRRTL] Add FIRParser support for classes

### PR DESCRIPTION
**Pull request by rwy7**
This PR adds support for parsing classes in FIRRTL.

### CODE REVIEW

**Review by mikeurbach — COMMENTED** [review]
Makes sense in general, though I'm not super familiar with this parser. The NFC refactoring can probably be merged directly first.

**Comment by rwy7** [comment]
refactor here: https://github.com/llvm/circt/pull/5710

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 1]
```diff
@@ -4193,6 +4195,26 @@ ParseResult FIRCircuitParser::parseParameterList(ArrayAttr &resultParameters) {
...
+      parseToken(FIRToken::colon, "expected ':' in class definition") ||
+      info.parseOptionalInfo() || parsePortList(portList, portLocs, indent))
+    return failure();
+  auto builder = circuit.getBodyBuilder();
+  auto classOp = builder.create<ClassOp>(info.getLoc(), name, portList);
+  deferredModules.emplace_back(
+      DeferredModuleToParse{classOp, portLocs, getLexer().getCursor(), indent});
+  if (skipToModuleEnd(indent))
+    return failure();
+  return success();
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:4215]: Only if you prefer, consider:
```suggestion
  return success(skipToModuleEnd(indent));
```

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 2]
```diff
@@ -4094,6 +4095,7 @@ ParseResult FIRCircuitParser::skipToModuleEnd(unsigned indent) {
     case FIRToken::kw_module:
     case FIRToken::kw_extmodule:
     case FIRToken::kw_intmodule:
+    case FIRToken::kw_class:
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:4098]: hyper-nit: Should we alpha-sort these?  The `parseFOO` methods appear to be organized that way.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 3]
```diff
@@ -4193,6 +4195,26 @@ ParseResult FIRCircuitParser::parseParameterList(ArrayAttr &resultParameters) {
...
+/// class ::= 'class' id ':' info? INDENT portlist simple_stmt_block DEDENT
+ParseResult FIRCircuitParser::parseClass(CircuitOp circuit, unsigned indent) {
+  StringAttr name;
+  SmallVector<PortInfo, 8> portList;
+  SmallVector<SMLoc> portLocs;
+  LocWithInfo info(getToken().getLoc(), this);
+  consumeToken(FIRToken::kw_class);
+  if (parseId(name, "expected class name") ||
+      parseToken(FIRToken::colon, "expected ':' in class definition") ||
+      info.parseOptionalInfo() || parsePortList(portList, portLocs, indent))
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:4207]: Should this be checking the ports for validity (only property-type)?

Verifier seems reasonable but checking in parser (or invoking verification logic somehow?) to produce nice diagnostics can be nice re:better error messages.

Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:4207]: Came to suggest an error test and giving these non-property ports was my first thing to try :smile: .

**Review by dtzSiFive — APPROVED** [review]
LGTM!

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp (hunk 2 above)
Comment (rwy7) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:4112]: alphabetized 👍

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp (hunk 3 above)
Comment (rwy7) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:4225]: added two simple checks below, and an error test for each. We still need to reject hardware ops under a class, but if it's okay, I will save that for a separate PR (it's not properly verified in the IR yet, either).

Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:4225]: Totally, sounds great! :+1:

**Pull request merged**